### PR TITLE
Openssl 3 0 update

### PR DESCRIPTION
**Pull request by ColeBollig**
-Tested Compilation in Fedora 36 docker container with openssl-devel (3.0.5) and saw no compilation warnings for use of deprecated functions
-Made to be compatible with JWT-CPP 0.6.0 for its openssl 3.0 support
-Minimal testing (Due to lack of know how). scitokens-verify works for both RSA and EC methods.

-Please scrutinize everything. I don't want to break scitokens

### CODE REVIEW

File: src/scitokens_internal.cpp [hunk 1]
```diff
@@ -243,8 +249,60 @@ std::string
...
+    std::unique_ptr<EVP_PKEY_CTX, decltype(&EVP_PKEY_CTX_free)> ec_ctx(EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL), EVP_PKEY_CTX_free);
+    if (!ec_ctx.get()) {
+        throw UnsupportedKeyException("Failed to set EC PKEY context");
+    }
+	
+	if (EVP_PKEY_fromdata_init(ec_ctx.get()) <= 0
+	    || EVP_PKEY_fromdata(ec_ctx.get(),&pkey,EVP_PKEY_PUBLIC_KEY,params) <= 0
+	    || pkey == NULL) {
+	        throw UnsupportedKeyException("Failed to set the EC public key");
+	}
```
Comment (djw8605) [inline src/scitokens_internal.cpp:296]: Indented oddly.  On purpose?

File: src/scitokens_internal.cpp [hunk 2]
```diff
@@ -284,29 +341,57 @@ std::string
...
+	    || EVP_PKEY_fromdata(rsa_ctx.get(),&pkey,EVP_PKEY_PUBLIC_KEY,params) <= 0
+	    || pkey == NULL) {
+	        throw UnsupportedKeyException("Failed to set the RSA public key");
+	}
+	
+    if (PEM_write_bio_PUBKEY(pubkey_bio.get(), pkey) == 0) {
+        throw UnsupportedKeyException("Failed to serialize RSA public key");
+    }
+    EVP_PKEY_free(pkey);
+	OSSL_PARAM_free(params);
```
Comment (djw8605) [inline src/scitokens_internal.cpp:374]: Odd indention

**Review by djw8605 — COMMENTED** [review]
Few minor comments.  I'll test this as well before merging.

File: src/scitokens_internal.cpp (hunk 2 above)
Comment (ColeBollig) [reply src/scitokens_internal.cpp:374]: That one was an auto tab from return due to my text editor. Will fix.

**Comment by djw8605** [comment]
Lots of errors in the build.  Could you take a look.

**Comment by ColeBollig** [comment]
I believe the build failures are due to the code requiring at minimum JWT-CPP v0.5.0 for the json traits change that breaks the api and isn't getting it. In theory you want v0.6.0 for the openssl 3.0 update within JWT-CPP. I am not sure how to change it so the github build uses the correct JWT-CPP version

**Comment by ColeBollig** [comment]
I have figured out how to make the base requirement of jwt-cpp

**Comment by ColeBollig** [comment]
Depending on the rpm building test and your approval I believe all changes are here. I do have a couple of things to note.
1. You should update the Unit test to test for both Openssl 3 and older Openssl versions.
2. This should resolve open issues [#72](https://github.com/scitokens/scitokens-cpp/issues/72) and [#61](https://github.com/scitokens/scitokens-cpp/issues/61)
3. Updated vendor copy of JWT-CPP is v0.6.0 with some added code from a later [JWT-CPP patch](https://github.com/Thalhammer/jwt-cpp/pull/246) that fixed rpm build issues. This should probably be replaced when an official version with the change is released.
4. I kept in the [code](https://github.com/scitokens/scitokens-cpp/commit/217a819e0ecd7697674712e45d864c014102a5f8) to turn off warnings becoming errors but made the default to ON. In case another deprecation warning issue occurs again.

Tested in docker Fedora:36 container with Openssl 3. scitoken-verify was successful for both RS256 and ES256 token from demo.scitokens.org, and successfully passed all unit tests.

**Comment by djw8605** [comment]
After a few iterations, I think I'm happy with the pull request.  Summary of my changes:
- Added newer versions of ubuntu with openssl 3.
- Completely removed the option to turn off warnings as errors
- Update the gtest submodule because newer compilers threw some errors

**Pull request merged**